res/controllers/midi-components-0.0.js

```diff
@@ -503,19 +503,25 @@
         }
     };
     ComponentContainer.prototype = {
-        forEachComponent: function (operation, recursive) {
+        forEachComponent: function (operation, recursive, including_containers) {
             if (typeof operation !== 'function') {
                 print('ERROR: ComponentContainer.forEachComponent requires a function argument');
                 return;
             }
             if (recursive === undefined) { recursive = true; }
+            if (including_containers === undefined) { including_containers = false; }
```

Comment:
Would there be a harm in making this true by default? Making it true by default would remove the need to explicitly specify the last 2 parameters in the Button.prototype.shift/unshift functions.

Reply:
> Just doing this unconditionally might break existing controller mappings since
containers do not inherit from Component and do not implement the same
methods.

Yeah, I think you are right. On further thought, this seems hacky to repurpose a function intended to be used on Components to also be used for ComponentContainers... maybe there is another solution?

Reply:
How about adding an optional parameter to `forEachComponent` that is a function that gets called on ComponentContainers (in addition to the recursive call to `forEachComponent`)? That way, the shift/unshift functions could pass simple functions like:
`function () { this.isShifted = true; }`
This would keep a separation of concerns between the functions that act on Components and those that act on ComponentContainers.

Reply:
> How about adding an optional parameter to forEachComponent that is a function that gets called on ComponentContainers (in addition to the recursive call to forEachComponent)?

Also feasible, but IMHO this is just as "hacky" as this solution. Also, this messes up the signature a bit `forEachComponent(func on_component, bool recusive, func on_container)`.

> On further thought, this seems hacky to repurpose a function intended to be used on Components to also be used for ComponentContainers... maybe there is another solution?

Another alternative might be to write a new method `forEachComponentContainer`. That would keep those concern separated, keep backwards compatibility and look cleaner overall. Downside: A few lines of duplicated code.
In `ComponentContainer.shift()`, we could then just set `recursive` to `false` and invoke `this.forEachComponentContainer(function(container) { container.shift(); })`.

Reply:
> Another alternative might be to write a new method forEachComponentContainer. That would keep those concern separated, keep backwards compatibility and look cleaner overall.

I like this idea.

Reply:
Ok, I reverted my original attempt and pushed a new implementation.